File: 67-add-binary/67-add-binary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        n, m, ans = len(a), len(b), []
        
        carry, i, j = 0, n - 1, m - 1
        while i >= 0 and j >= 0:
            carry, d = divmod(int(a[i]) + int(b[j]) + carry, 2)
            ans.append(str(d))
            i -= 1
            j -= 1
        
        k, c = (i, a) if i >= 0 else (j, b)
        while k >= 0:
            carry, d = divmod(int(c[k]) + carry, 2)
            ans.append(str(d))
            k -= 1
        
        if carry: ans.append("1")
            
        return "".join(reversed(ans))
        
        

File: 67-add-binary/67-add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        if(a.length() < b.length()){
            String t = a;
            a = b;
            b = t;
        }
        
        StringBuilder sb = new StringBuilder();
        int carry = 0;
        int j = b.length() - 1;
        for(int i = a.length() - 1; i >= 0; --i){
            int sum = a.charAt(i) - '0' + carry;
            if(j >= 0){
                sum += b.charAt(j) - '0';
                --j;
            }
            
            carry = sum / 2;
            sb.append((char) (sum % 2 + 48));
        }
        
        if(carry > 0){
            sb.append('1');
        }
        
        return sb.reverse().toString();
    }
}